packages/solana-intents/src/lib.rs

```diff
@@ -109,7 +107,52 @@ impl Ed25519InstructionHeader {
             message_instruction_index: u16::MAX,
             message_data_size: self.message_data_size,
         };
-        self == &expected_header
+        if self == &expected_header {
+            Ok(())
+        } else {
+            Err(IntentError::SignatureVerificationUnexpectedHeader)
+        }
+    }
+}
+
+enum Message {
+    Raw(Vec<u8>),
+    OffchainMessage(OffchainMessage),
+}
+
+impl From<Message> for Vec<u8> {
+    fn from(message: Message) -> Self {
+        match message {
+            Message::Raw(message) => message,
+            Message::OffchainMessage(message) => message.get_message().to_vec(),
+        }
+    }
+}
+
+fn get_length_with_header(message: &OffchainMessage) -> usize {
+    match message {
+        OffchainMessage::V0(_) => message.get_message().len() + OffchainMessage::HEADER_LEN + 3,
+    }
+}
+
+impl Message {
+    fn deserialize(data: &[u8]) -> std::io::Result<Self> {
+        if OffchainMessage::SIGNING_DOMAIN.len() <= data.len()
+            && data[0..OffchainMessage::SIGNING_DOMAIN.len()] == *OffchainMessage::SIGNING_DOMAIN
+        {
+            let message = OffchainMessage::deserialize(data).map_err(|_| {
+                std::io::Error::new(std::io::ErrorKind::InvalidData, "Invalid offchain message")
```

Comment:
`std::io::Error` feels wrong here -- this is not IO per se, the IO has already happened at this point